gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -173,3 +173,20 @@ def sqrt_inv_matmul(self, rhs, lhs=None):
     def zero_mean_mvn_samples(self, num_samples):
         base_samples = torch.randn(num_samples, *self._diag.shape, dtype=self.dtype, device=self.device)
         return base_samples * self._diag.sqrt()
+
+
+class ConstantDiagLazyTensor(DiagLazyTensor):
+    def __init__(self, diag_values, diag_shape):
+        """
+        Diagonal lazy tensor with constant entries. Supports arbitrary batch sizes.
+        Used e.g. for adding jitter to matrices.
+
+        Args:
+            :attr:`n` (int):
+                The (non-batch) dimension of the (square) matrix
+            :attr:`diag_values` (Tensor):
+                A `b1 x ... x bk x 1` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` diagonal matrices
+        """
+        super(DiagLazyTensor, self).__init__(diag_values, diag_shape=diag_shape)
+        self._diag = diag_values.expand(*diag_values.shape[:-1], diag_shape)
```

Comment:
Ends up not being a super difficult fix. This gets all but some strange missing `_quad_form_derivative` errors in `test_mul_lazy_tensor.py`

```suggestion
        self._diag = diag_values.expand(*diag_values.shape[:-1], diag_shape)
        
    def _expand_batch(self, batch_shape):
        return self.__class__(self._diag.expand(*batch_shape, self._diag.size(-1)), diag_shape=self.diag_shape)
    
    def _sum_batch(self, dim):
        return self.__class__(self._diag.sum(dim), diag_shape=self.diag_shape)
```

Reply:
Thanks for fixing this, updated. The test failures are weird, here's what's happening: Looks like the custom derivative still spits out  a gradient for the third arg while the auto (`gpytorch.lazy.LazyTensor._quad_form_derivative`) does not. This is more a thing for @jacobrgardner than for me...

```
> /Users/balandat/Code/gpytorch/gpytorch/test/base_test_case.py(26)assertAllClose()
     24 
     25     def assertAllClose(self, tensor1, tensor2, rtol=1e-4, atol=1e-5, equal_nan=False):
---> 26         if not tensor1.shape == tensor2.shape:
     27             raise ValueError(f"tensor1 ({tensor1.shape}) and tensor2 ({tensor2.shape}) do not have the same shape.")
     28 

ipdb> up
> /Users/balandat/Code/gpytorch/gpytorch/test/lazy_tensor_test_case.py(240)test_quad_form_derivative()
    238 
    239         for dc, da in zip(deriv_custom, deriv_auto):
--> 240             self.assertAllClose(dc, da)
    241 
    242     def test_sum(self):

ipdb> dc
tensor([ 0.6897, -1.8754, -1.2027,  2.2910, -4.5156, -1.2208])
ipdb> da
ipdb> deriv_custom
(tensor([[  12.5706,   -2.7285,    4.4291,  -17.3775,   -6.0653,   -0.2427],
        [  46.8123,  -16.0242,  -27.8904,   52.4139,   27.1451,    5.6616],
        [  20.9721,  -14.7426,  -27.9800,  -23.7657,   -8.3122,    1.3579],
        [   5.5446,  -25.0559,  -14.1486,   98.4952,   -5.6822,   -6.5448],
        [  17.0694,  -53.7562,   28.4515, -150.3875,  -64.8544,    6.9690],
        [ -34.1945,   -6.0813,    3.3062,   11.1382,  -14.7538,  -15.7924]],
       grad_fn=<AddBackward0>), tensor([[ 14.3406,  17.7023,   4.9514,  -2.1083,  -6.1771,  -2.8885],
        [ 15.2344, -18.4544,  -7.5149, -17.3055,  -9.3671,  -0.2566],
        [  6.3595, -12.7781, -59.6444, -17.0256,  -8.6747,   5.2716],
        [ 19.8825,  27.7598,   6.0443,  37.0225,  -7.2952,  -0.5087],
        [ 35.4283,  75.5914,   6.9602, -77.7236, -61.9332,   2.3141],
        [-11.2108,  23.8441,  14.2643,  39.6170,  -5.5581, -13.7373]],
       grad_fn=<AddBackward0>), tensor([ 0.6897, -1.8754, -1.2027,  2.2910, -4.5156, -1.2208]))
ipdb> deriv_auto
(tensor([[  12.5706,   -2.7285,    4.4291,  -17.3775,   -6.0653,   -0.2427],
        [  46.8123,  -16.0242,  -27.8904,   52.4139,   27.1451,    5.6616],
        [  20.9721,  -14.7426,  -27.9800,  -23.7657,   -8.3122,    1.3579],
        [   5.5446,  -25.0559,  -14.1486,   98.4952,   -5.6822,   -6.5448],
        [  17.0694,  -53.7562,   28.4515, -150.3875,  -64.8544,    6.9690],
        [ -34.1945,   -6.0813,    3.3062,   11.1382,  -14.7538,  -15.7924]]), tensor([[ 14.3406,  17.7023,   4.9514,  -2.1083,  -6.1771,  -2.8885],
        [ 15.2344, -18.4544,  -7.5149, -17.3055,  -9.3671,  -0.2566],
        [  6.3595, -12.7781, -59.6444, -17.0256,  -8.6747,   5.2716],
        [ 19.8825,  27.7598,   6.0443,  37.0225,  -7.2952,  -0.5087],
        [ 35.4283,  75.5914,   6.9602, -77.7236, -61.9332,   2.3141],
        [-11.2108,  23.8441,  14.2643,  39.6170,  -5.5581, -13.7373]]), None)
```

Reply:
That's what I was seeing last night. I wonder if we ought to be storing the `diag` value and expanding it properly to remove any gradient wackiness --- taking a look now.